Begin paths for internationalization in kubectl

brendandburns opened the pull request:
This is just the first step, purposely simple so we can get the interface correct.

@kubernetes/sig-cli @deads2k

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36802)
<!-- Reviewable:end -->

deads2k commented:
@smarterclayton @bgrant0607 @kubernetes/sig-api-machinery This would be a very large, rippling change across the entire code base.  Stopping at the CLI won't help much since most useful errors come from the server.

deads2k commented on pkg/util/i18n/translations/en-us.go:
```diff
@@ -0,0 +1,16 @@
+
+var EnUSFileName = "en-US.all.json"
+
+var EnUSData = []byte(`
```
Is this really the best format available to us?  Java style resource bundle properties files seem less fragile.

fabianofranz replied:
Yes. Using a better-established format for the translation files favours translators that are not programmers and a world of tools that exist to support them. JSON and YAML are usually acceptable but something closer to plain text is usually better, like Java or Mozilla .properties, CSVs like `id,translated text` and so on. Conversions scripts would probably also work, though.

brendandburns replied:
I inlined this JSON because I was hoping to avoid yet more gobindata, but I guess we'll have to go there...

I'm not opposed to conversion, is there a generally accepted format?  Making one up just so that we don't use JSON doesn't seem worth it...

deads2k commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -107,7 +109,7 @@ func NewCmdAnnotate(f cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:     "annotate [--overwrite] (-f FILENAME | TYPE NAME) KEY_1=VAL_1 ... KEY_N=VAL_N [--resource-version=version]",
-		Short:   "Update the annotations on a resource",
+		Short:   T("annotate_short"),
```
If we do this, I think the pattern should be constants for all translated string keys.  One constant file per resource bundle with a static analyzer to catch mismatches.  I've had better luck with that.

fabianofranz replied:
Writing code by having to manually put all your messages in a separate file, defining an id, etc is painful.

What about something like

```
cmd := &cobra.Command{
  (...)
  Short:   T("Update the annotations on a resource"),
```

Then we use `go/ast` to write a smartass Go binary that gets all `CallExpr` of `T(...)` and generate the default (en-US) messages file for us?

```
pkg/kubectl/cmd/annotate#update-the-annotations-on-a-resource: Update the annotations on a resource
pkg/kubectl/cmd/annotate#duplicate-message: A duplicate message
pkg/kubectl/cmd/annotate#duplicate-message2: Another duplicate message
...
```

The message keys can obviously be anything we decide, just has to be uniquely auto-generated given what we have (file name, the actual message string, etc).

And even better, you don't even need the generated file because if a messages file for the wanted language is not present, you just default for what `T` takes as an argument... messages file generation can be used only as a stub for translating the other languages. And while you don't do translations you don't need messages files at all.

deads2k replied:
> And even better, you don't even need the generated file because if a messages file for the wanted language is not present, you just default for what T takes as an argument... messages file generation can be used only as a stub for translating the other languages.

Don't we need stable keys to be able to locate the translated string?  That would seem to require putting some identifier in the code to keep the key stable for file moves and the like.  It's also not obvious how this approach would deal with, "I want the same message as that guy over there".

fabianofranz replied:
> Don't we need stable keys to be able to locate the translated string? 

A key only have to be "stable" while its message doesn't change. If the message changes it has to be translated again, right? So it becomes a new message anyway, with a new key. 

That said, you generate the keys, as in 

```
T("Update the annotations on a resource")
```

becomes for example the key

```
kubectl-annotate#update-the-annotations-on-a-resource
```

`T` takes a message string, generate the key from it and search for that key in the translated file for the language you want. If not found, just use the default message, which is just `T` argument.

> It's also not obvious how this approach would deal with, "I want the same message as that guy over there".

I don't recommend reusing messages, even if they are as simple as a single word. Every single call to `T` must relate directly to one dedicated message key. Exception being maybe inside the same command. But duplicated strings in translation files are ok and expected.

fabianofranz replied:
Still, if none of that sound interesting, one intermediate idea would be

```
T("annotate_short", "Update the annotations on a resource")
```

Eliminates the need of a separate file for the default language which addresses "writing code by having to manually put all your messages in a separate file is painful" and makes the code more readable.

deads2k commented on vendor/github.com/nicksnyder/go-i18n/i18n/language/plural.go:
```diff
@@ -0,0 +1,40 @@
+
+// Plural represents a language pluralization form as defined here:
+// http://cldr.unicode.org/index/cldr-spec/plural-rules
+type Plural string
```
If this is done, you should make an example with pluralization rules.

smarterclayton commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -31,6 +31,7 @@ import (
 	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
 	"k8s.io/kubernetes/pkg/kubectl/resource"
 	"k8s.io/kubernetes/pkg/runtime"
+	. "k8s.io/kubernetes/pkg/util/i18n"
```
I would prefer not to do this, `i18n.T` is short and it plays much better with tools.

smarterclayton commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -92,6 +93,7 @@ var (
 )
 
 func NewCmdAnnotate(f cmdutil.Factory, out io.Writer) *cobra.Command {
+	LoadTranslations()
```
If you do it this way we will have to load all translation files from disk for every single command ever run, even if it uses no actual translation output.

smarterclayton replied:
Actually, I take that back.  Since the eventual impl is probably a properties file compiled into the binaries for most code paths init is much cheaper.  Although I wouldn't want to decorate all the command code paths with init code.  Maybe lazy is better there, or just an init early in code.

I don't think single bundle for binary is going to work for hyperkube, so we're going to need packages per bundle type.

eparis replied:
I'm sure at least one cobra maintainer would be interested to know what was possible...

fabianofranz commented:
Have you considered just using the packages behind `golang.org/x/text` instead of `github.com/nicksnyder/go-i18n`, like https://godoc.org/golang.org/x/text/message and https://godoc.org/golang.org/x/text/internal/format/plural? (I'm not arguing in favour or against, just want to understand).

fabianofranz commented:
@kubernetes/kubectl

dims commented:
based on @fabianofranz 's comment on being translator friendly..

openstack uses zanata as zanata is very translator friendly
https://wiki.openstack.org/wiki/Translations
https://translate.openstack.org/

and the PO files could be loaded using gettext golang bindings

fabianofranz commented on pkg/util/i18n/i18n.go:
```diff
@@ -0,0 +1,68 @@
+
+// LoadTranslations loads translation files.
+func LoadTranslations() error {
+	lang := os.Getenv("LANG")
```
Need to check how this goes on Windows / Mac.

smarterclayton commented:
A lot of translation libraries just key off the actual string.  That has a
lot of advantages (that's how a large slice of applications I've I18n'ized
over the last 15 years have done it).

On Fri, Nov 18, 2016 at 6:27 AM, Fabiano Franz <notifications@github.com>
wrote:

> *@fabianofranz* commented on this pull request.
> ------------------------------
>
> In pkg/kubectl/cmd/annotate.go
> <https://github.com/kubernetes/kubernetes/pull/36802>:
>
> > @@ -107,7 +109,7 @@ func NewCmdAnnotate(f cmdutil.Factory, out io.Writer) *cobra.Command {
>
>  	cmd := &cobra.Command{
>  		Use:     "annotate [--overwrite] (-f FILENAME | TYPE NAME) KEY_1=VAL_1 ... KEY_N=VAL_N [--resource-version=version]",
> -		Short:   "Update the annotations on a resource",
> +		Short:   T("annotate_short"),
>
> Still, if none of that sound interesting, one intermediate idea would be
>
> T("annotate_short", "Update the annotations on a resource")
>
> Eliminates the need of a separate file for the default language which
> addresses "writing code by having to manually put all your messages in a
> separate file is painful" and makes the code more readable.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36802>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pxOUC2tieYtfhfPY_aNKpmUaoh6Kks5q_bW4gaJpZM4KyOKS>
> .
>

brendandburns commented:
Ok, I moved to just using the string.

I switched to an external YAML file instead of the inline go JSON

I added a pluralization example (and tests)

Please take another look.

Thanks!
--brendan

lavalamp commented on pkg/kubectl/cmd/get.go:
```diff
@@ -217,7 +218,7 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 		}
 		if len(infos) != 1 {
-			return fmt.Errorf("watch is only supported on individual resources and resource collections - %d resources were found", len(infos))
+			return fmt.Errorf(i18n.T("watch is only supported on individual resources and resource collections - {{.Count}} resources were found", len(infos)))
```
Need to switch to just plain NewError to avoid double-interpreting strings. (This one doesn't matter but others will and people will copy-pasta this.)

lavalamp replied:
Better yet, just provide a i18n.Errorf function. (or i18n/errors library.)

lavalamp replied:
I'm a little concerned that people will change the string and the yaml file will be broken, is that condition detected?

brendandburns replied:
Done (except for the detect part), the default is return the same string, e.g. no translation, which seems like a sane default.  Later on we can add a script to check for untranslated strings (which we're going to need to do anyway)

lavalamp commented on pkg/generated/BUILD:
```diff
@@ -12,6 +12,9 @@ load(
     name = "go_default_library",
-    srcs = ["doc.go"],
+    srcs = [
+        "bindata.go",
```
Maybe this was already discussed but is it REALLY necessary to compile this in?

brendandburns replied:
There are two reasons for compiling in:
1) Load time performance so that we don't slow down kubectl by loading lots of files.
2) Packaging, it's really good that right now you can ship `kubectl` and nothing else and it just works.  Translations should be no different.

(and hey `kubectl` is already > 70Mb)

lavalamp commented on pkg/util/i18n/i18n.go:
```diff
@@ -0,0 +1,113 @@
+
+var (
+	lang language.Tag
+	bdl  *bundle.Bundle
```
Can we avoid making this a global and/or is there some other way we can avoid making our unit tests depend on the system language? Otherwise we have a situation where adding a Polish translation breaks everyone in WAW...

lavalamp replied:
So how is this going to work when:

1. kubectl, apiserver, controller-manager etc live in separate repositories
2. all the above are compiled into hyperkube

lavalamp commented on pkg/util/i18n/i18n.go:
```diff
@@ -0,0 +1,113 @@
+	lock sync.Mutex
+)
+
+func init() {
```
Like, maybe don't call this from init, make the binary call this to enable it.

brendandburns replied:
done.

lavalamp commented on pkg/util/i18n/i18n.go:
```diff
@@ -0,0 +1,113 @@
+	return nil
+}
+
+func T(defaultValue string, args ...interface{}) string {
```
Document. Does this use %v formatters or do you have to switch to {{}}?

lavalamp replied:
Document the plural behavior, or say where to go to read about it.

brendandburns replied:
Done.

lavalamp commented on translations/en-US.all.yaml:
```diff
@@ -0,0 +1,17 @@
+
+- id: "watch is only supported on individual resources and resource collections - {{.Count}} resources were found"
+  translation:
+    one: "watch is only supported on individual resources and resource collections - {{.Count}} resources was found"
```
How does it know to use the value of {{.Count}} to switch between one and other? and shouldn't it be "1" and other? And is this really the way we want to do plurals?

(Also should be s/resources was/resource was/)

brendandburns replied:
This is part of the `github.com/nicksnyder/go-i18n` package.  Added links to examples.

(and fixed the string)

I'm open to other suggestions for libraries for doing translation/pluralization.  `go-i18n` was the best I could find.

lavalamp commented on pkg/util/i18n/i18n.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// LoadTranslations loads translation files.
+func LoadTranslations() error {
```
Do we really need to build this? I'm very surprised that something like this doesn't already exist.

lavalamp commented:
So I think we need to plan for this to be used/loaded at smaller amounts of granularity than entire binaries. That lets us roll it out piecemeal and also compile various bits into various combination binaries.

brendandburns commented:
Comments largely addressed, please re-check.

Thanks

lavalamp commented:
This is looking a lot better, thanks.

I would like to know what your plan is for converting strings to go through the i18n library, for actually getting translations, which components you want to update, and how (or if) you're going to get a user's language plumbed through apiserver to controller manager where many errors are generated.

tagoh commented:
Can we use the gettext-based format to store the translations or at least friendly-format to be converted from? most translators are familiar with it or translation platforms is supporting. using another one takes the extra costs to support i18n/l10n.

brendandburns commented:
@smarterclayton @fabianofranz can you comment on the relative value/ease of `gettext` formatted input vs. other formats?

There is a couple of golang `gettext` libraries (e.g. https://github.com/chai2010/gettext-go)

So I could refactor this code in favor of that format fairly easily...

Let me know what you think.

--brendan

fabianofranz commented:
> can you comment on the relative value/ease of gettext formatted input vs. other formats?

Either `gettext` (extension is usually `.po`) or Java/Mozilla `.properties` are our two best choices. 

They are closer to plain text that YAML/JSON and others, and are widely accepted by translators and translation tools. Check for example [Matecat](https://www.matecat.com/), [Google Translator Toolkit](https://translate.google.com/toolkit/docupload?hl=pt_BR), and [Zanata](http://zanata.org/features/), all of them support it. No strong preference between the two, so in the end, I'd say it's a matter of which of them has the best Go library.

Disclaimer: wife is a translator.  ;)

jwforres commented:
I don't know what dashboard is planning ( @bryk ? ) but for openshift
console we will most likely use gettext, there are nice libraries for
angular.  It would be nice if the translators didn't have to work in
different formats for different clients.

On Mon, Dec 5, 2016 at 12:30 PM, Fabiano Franz <notifications@github.com>
wrote:

> can you comment on the relative value/ease of gettext formatted input vs.
> other formats?
>
> Either gettext (extension is usually .po) or Java/Mozilla .properties are
> our two best choices.
>
> They are closer to plain text that YAML/JSON and others, and are widely
> accepted by translators and translation tools. Check for example Matecat
> <https://www.matecat.com/>, Google Translator Toolkit
> <https://translate.google.com/toolkit/docupload?hl=pt_BR>, and Zanata
> <http://zanata.org/features/>, all of them support it. No strong
> preference between the two, so in the end, I'd say it's a matter of which
> of them has the best Go library.
>
> Disclaimer: wife is a translator. ;)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36802#issuecomment-264918922>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7RbtdT8TViiBcyyG6HUQ5lazlCWhks5rFEoZgaJpZM4KyOKS>
> .
>

bryk commented:
> I don't know what dashboard is planning ( @bryk ? ) but for openshift
console we will most likely use gettext, there are nice libraries for
angular.  It would be nice if the translators didn't have to work in
different formats for different clients.

Dashboard UI is already i18n-ed. We translate to Japanese and English. We use XTB file format, but we've been already thinking about switching to gettext. 

@jwforres We've created a developer-friendly and performant framework for translating angular apps. You could reuse it in openshift. 

cc @kubernetes/dashboard-maintainers

brendandburns commented:
Ok, I will switch this to gettext format.

Thanks all.

brendandburns committed:
Add github.com/chai2010/gettext-go dependency

brendandburns committed:
Add initial translation support.

brendandburns commented:
Rebased and switched to gettext formatted files.  Please take another look.

Thanks!
--brendan

brendandburns commented:
@fabianofranz @deads2k friendly ping on this one.

Thanks
-brendan

fabianofranz commented on pkg/util/i18n/i18n.go:
```diff
@@ -0,0 +1,124 @@
+	}
+	gettext.BindTextdomain("k8s", root+".zip", buf.Bytes())
+	gettext.Textdomain("k8s")
+	gettext.SetLocale(langStr)
```
Is this safe about setting a `langStr` for which you don't have a translation?

fabianofranz commented on translations/test/en_US/LC_MESSAGES/k8s.po:
```diff
@@ -0,0 +1,30 @@
+
+msgctxt "test_string"
+msgid "test_string"
+msgstr "baz"
```
Do you need files for both `default` and `en_US` languages? Why is this message different between the two?

brendandburns replied:
Yeah, I want this for the unit test, so I can try explicit language loading instead of the default.

The reason it is different so that the unit test won't pass if we are accidentally loading the "default" language for some reason.

I want `translate("default", "test_string") != translate("en_US, "test_string")` otherwise a unit test might accidentally 'pass' even though the code is actually wrong.

fabianofranz commented:
Looking pretty good. I'd like one more comments but LGTM as the interface for translations. As a follow-up we could consider an actual integration of the messages in the repository with a Zanata server, which gives you a nice dashboard for messages versions and completeness for each language, etc. Example: https://translate.jboss.org/iteration/view/dashbuilder/6.5.0?dswid=8235.

brendandburns commented:
@fabianofranz comments addressed, please take another look.

Thanks!
--brendan

fabianofranz commented on cmd/kubectl/app/kubectl.go:
```diff
@@ -33,6 +34,12 @@ Any salient changes here will need to be manually reflected in that file.
 func Run() error {
 	logs.InitLogs()
 	defer logs.FlushLogs()
+	i18n.LoadTranslations("kubectl", nil, []string{
```
Can be a follow-up, but since you only care about the set of messages for one single language at each `kubectl` run, the language detection could happen here and you'd only load the language you care about.

brendandburns replied:
done. (moved to the LoadTranslations function)

fabianofranz commented on pkg/util/i18n/i18n.go:
```diff
@@ -0,0 +1,124 @@
+// string (e.g. 'en-US'). If getLanguageFn is nil, then the loadSystemLanguage function
+// is used. translationFiles are paths to translation files that are expected to be
+// compiled into the binary via go-bindata.
+func LoadTranslations(root string, getLanguageFn func() string, translationFiles []string) error {
```
In future we'll want a `getLanguageFn` that loads from preferences or something similar, so that I can manually set it.

brendandburns replied:
By default getLanguageFn loads from the `LANG` environment variable, which seems like a decent way to set this preference.  Added a TODO for a flag, but it happens so early in the startup, I'm not even sure how to plumb a flag through to that point.

fabianofranz commented:
LGTM, a couple comments that can be follow-ups.

brendandburns commented:
Comments addressed.  Will self-LGTM on your LGTM once it's green.

fabianofranz commented:
@k8s-bot verify test this
